### PR TITLE
rpc: don't serialize a full `Job` during plan submission

### LINKED ISSUE #23766: plan submit should not send serialized job version
When a scheduler submits a plan to the leader, the `Plan.Submit` RPC includes the entire `Job` object from the state store of the server where the scheduler ran. When originally written there were no deployments or job version tracking, so the scheduler had to assume the job had been changed out from under it. 

But this makes safely submitting plans during a cluster upgrade much more complicated. We canonicalize a job when we upsert it in the state store via Raft (ref [`fsm.go#L607-L614`](https://github.com/hashicorp/nomad/blob/v1.8.2/nomad/fsm.go#L607-L614)), and when we restore from snapshot (ref [`fsm.go#L1587`](https://github.com/hashicorp/nomad/blob/v1.8.2/nomad/fsm.go#L1587)), which happens when you upgrade a server. But a scheduler running on an older version of Nomad can submit a plan with a job that has not been canonicalized (so potentially including unexpected nil pointers), and the plan applier has to operator correctly on that job.

We could instead have the plan include the `(namespace, job_id, version)` tuple, and then the leader would read the job from its own state store. This would reduce the size of the `Plan.Submit` request and improve safety across cluster upgrades. The plan applier could reject plans for versions earlier than the latest, and that would prevent applying a plan that was generated from an earlier version of the job. We'd only have to worry about drift of updates to the same version, which are minimal (stopping, scaling, etc.) and for which we already have logic to repair by sending new evals.

The upgrade path for this change would itself be a little hairy: we'd need to ensure we're still sending the job with the plan if the leader isn't of a minimum Nomad version with this change. Another option to consider is gating the change behind a [scheduler configuration flag](https://developer.hashicorp.com/nomad/docs/commands/operator/scheduler/set-config).

### PR DESCRIPTION
**Pull request by pkazmierczak**
There's no need to submit a full `Job` object during `Plan.Submit` RPC, since servers can pull the job from their state. 

Resolves #23766

### CODE REVIEW

File: nomad/structs/plan.go [hunk 1]
```diff
@@ -36,6 +36,11 @@ type Plan struct {
 	// of the plan by only including it once.
 	Job *Job
 
+	// JobInfo contains namespace, job ID and version of all the allocations
+	// in the Plan. This is so that we don't serialize the whole Job object in the
+	// Plan.Submit RPC.
```
Comment (tgross) [inline nomad/structs/plan.go]: Nitpicking the docstring a bit: not all the allocations in the Plan will match this tuple, because we can have allocations from previous versions that we're stopping.

File: nomad/structs/plan.go [hunk 2]
```diff
@@ -138,9 +155,14 @@ func (p *Plan) AppendStoppedAlloc(alloc *Allocation, desiredDesc, clientStatus,
...
-	// extract the job from the allocation.
-	if p.Job == nil && newAlloc.Job != nil {
+	// extract the job information from the allocation.
+	if (p.Job == nil && p.JobInfo == nil) && newAlloc.Job != nil {
 		p.Job = newAlloc.Job
+		p.JobInfo = &PlanJobTuple{
+			Namespace: newAlloc.Job.Namespace,
+			ID:        newAlloc.Job.ID,
+			Version:   newAlloc.Job.Version,
+		}
```
Comment (tgross) [inline nomad/structs/plan.go:162]: Doesn't this overwrite the plan's info with the old allocation?

File: nomad/plan_apply_test.go [hunk 3]
```diff
@@ -1106,7 +1181,11 @@ func TestPlanApply_EvalNodePlan_NodeDown_EvictOnly(t *testing.T) {
 	*allocEvict = *alloc
 	allocEvict.DesiredStatus = structs.AllocDesiredStatusEvict
 	plan := &structs.Plan{
-		Job: alloc.Job,
+		JobInfo: &structs.PlanJobTuple{
+			Namespace: alloc.Job.Namespace,
+			ID:        alloc.Job.ID,
+			Version:   alloc.Job.Version,
+		},
```
Comment (tgross) [inline nomad/plan_apply_test.go:1179]: This test case points out a weird edge case I hadn't considered in the original issue: if the plan _only_ contains older allocations being stopped, which version should we use?

File: nomad/worker.go [hunk 4]
```diff
@@ -656,6 +656,18 @@ func (w *Worker) SubmitPlan(plan *structs.Plan) (*structs.PlanResult, sstructs.S
 	}
 	defer metrics.MeasureSince([]string{"nomad", "worker", "submit_plan"}, time.Now())
 
+	// Figure out whether we need to submit a plan that contains a full job or a
+	// new, "lean" plan with just the basic job info (ns, id and ver)
+	if w.ServersMeetMinimumVersion(minVersionPlanLeanJob, false) {
+		job := plan.Job.Copy()
```
Comment (tgross) [inline nomad/worker.go]: I don't think we need this copy -- all the fields we're assigning to `JobInfo` are primitives.

File: nomad/structs/plan.go (hunk 2 above)
Comment (pkazmierczak) [reply nomad/structs/plan.go:162]: No. The conditional checks if `p.Job == nil && p.JobInfo == nil`.

**Comment by pkazmierczak** [comment]
> Overall this looks good but I think the implementation has raised questions about the value/purpose of having the version in the tuple. Are we actually consuming that anywhere?

I think you're right that we don't need version. I removed it in 52acc47
